fourm/data/modality_transforms.py

```diff
@@ -867,8 +947,129 @@ def postprocess(self, bboxes):
+
+        return bbs_per_frame
+
+    def preprocess(self, sample: List[dict]) -> List[List[tuple]]:
+        """
+        Given a list of dicts (each element in the list is a frame, each dict is a representation of bounding boxes for that frame),
+        convert the instances within each frame's bounding boxes representation.
+        # TODO: any issue with returning a list of lists here? We sorta might break the type consistency of the class? I'm not sure how robust it is.
+            # ^ Actually I think this is fine because it just gets passed into the image_augment function, so as long as that's implemented correspondingly we're good.
+
+        Returns:
+            List of lists of tuples.
+            The outermost list represents each frame.
+            The middle list represents each bounding box instances in that frame.
+            The innermost tuple is the actual representation of a bounding box instance (of the form [xmin, ymin, xmax, ymax, class_name, score]).
+        """
+        instances = [frame["instances"] for frame in sample]
+        return self.convert_detection_instance(instances)
+
+    def image_augment(
+        self,
+        bboxes_per_frame: List[List[Tuple]],
+        crop_coords: Tuple,
+        flip: bool,
+        orig_size: Tuple,
+        target_size: Tuple,
+        rand_aug_idx=None,
+        resample_mode: str = None,
+    ) -> List[List[Tuple]]:
+        """
+        Apply bbox augmentations to the bounding boxes of each frame (TODO: do we actually wanna do this? Would some of these augmentations break the video nature/spatio-temporal relations/etc.?)
+        """
+        bboxes_per_frame = []
+        for bboxes in bboxes_per_frame:
+            bboxes = self.bboxes_crop_and_resize(bboxes, crop_coords, orig_size)
+            bboxes = self.bboxes_hflip(bboxes, target_size, flip)
+            bboxes = self.order_and_filter_bboxes(bboxes)
+
+            bboxes_per_frame.append(bboxes)
+
+        return bboxes_per_frame
 
+    def postprocess(self, bboxes_per_frame):
+        """
+        Given bounding box representations per frame, should return a string like:
+        "<frame_0_token> bbox0 string representation blah blah blah <bbox_1> ... <frame_1_token>bbox0 string representation blah blah blah<bbox_1> ... <eos_token>"
+        """
+        if self.return_raw:
+            raise NotImplementedError(
+                "I'm not sure what the correct behavior of returning bboxes_per_frame should be when returned raw, so throwing error for now. Set return_raw=False to avoid this."
+            )
+            return bboxes_per_frame
+        output_str = "<frame_0_token>"  # TODO: implement the <frame_i_token> special tokens better than just hardcoding things here.
+        for i, bboxes in enumerate(bboxes_per_frame):
+            bboxes_str = self.convert_bboxes_to_string(bboxes_per_frame)
+            output_str += bboxes_str
+            output_str += f"<frame_{i+1}_token>"
```

Comment:
these new special tokens need to be added to the tokenizer jsons -- https://github.com/swiss-ai/ml-4m/blob/main/fourm/utils/tokenizer/trained/text_tokenizer_4m_wordpiece_30k.json (we can manually add these tokens to the json file)